pep-0677.rst

```diff
@@ -0,0 +1,851 @@
+of the AST and Grammar with tests verifying that the grammar proposed
+here has the desired behaviors.
+
+The runtime behavior is not yet implemented. As discussed in the
+`Runtime Behavior`_ portion of the spec we have a detailed plan for
+both a backward-compatible API and a more structured API in
+ `a separate doc
+<https://docs.google.com/document/d/15nmTDA_39Lo-EULQQwdwYx_Q1IYX4dD5WPnHbFG71Lk/edit>`_
+where we are also open to discussion and alternative ideas.
+
+
+Open Issues
+===========
+
+Details of the Runtime API
+--------------------------
+
+Once we have finalized all details of the runtime behavior, we
+will need to add a full specification of the behavior to the
+`Runtime Behavior`_ section of this PEP as well as include that
+behavior in our reference implementation.
+
+Optimizing ``SyntaxError`` messages
+-----------------------------------
+
+The current reference implementation has a fully-functional parser and
+all edge cases presented here have been tested.
+
+But there are some known cases where the errors are not as informative
+as we would like. For example, because ``(int, ...) -> bool`` is
+illegal but ``(int, ...)`` is a valid tuple, we currently produce a
+syntax error flagging the ``->`` as the problem even though the real
+cause of the error is using ``...`` as an argument type.
+
+This is not part of the specification *per se* but is an important
+detail to address in our implementation. The solution will likely
+involve adding ``invalid_.*`` rules to ``python.gram`` and customizing
+error messages.
+
+Resources
+=========
+
+Background and History
+----------------------
+
+`PEP 484 specifies
+<https://www.python.org/dev/peps/pep-0484/#suggested-syntax-for-python-2-7-and-straddling-code>`_
+a very similar syntax for function type hint *comments* for use in
+code that needs to work on Python 2.7. For example::
+
+    def f(x, y):
+        # type: (int, str) -> bool
+        ...
+
+At that time we used indexing operations to specify generic types like
+``typing.Callable`` because we decided not to add syntax for
+types. However, we have since begun to do so, e.g. with PEP 604.
+
+**Maggie** proposed better callable type syntax as part of a larger
+`presentation on typing simplifications
+<https://drive.google.com/file/d/1XhqTKoO6RHtz7zXqW5Wgq9nzaEz9TXjI/view>`_
+at the PyCon Typing Summit 2021.
+
+**Steven** `brought up this proposal on typing-sig
+<https://mail.python.org/archives/list/typing-sig@python.org/thread/3JNXLYH5VFPBNIVKT6FFBVVFCZO4GFR2>`. We
+had several meetings to discuss alternatives, and `this presentation
+<https://www.dropbox.com/s/sshgtr4p30cs0vc/Python%20Callable%20Syntax%20Proposals.pdf?dl=0>`_
+led us to the current proposal.
+
+**Pradeep** `brought this proposal to python-dev
+<https://mail.python.org/archives/list/python-dev@python.org/thread/VBHJOS3LOXGVU6I4FABM6DKHH65GGCUB>`_
+for feedback.
+
+The proposed AST and runtime behavior is explicitly designed to be
```

Comment:
This paragraph feels out of place; there's already discussion of forward compatibility above.